keycloak-claims-mapper/src/main/java/ee/eki/tolkevarav/sso/keycloakserviceprovider/tokenclaimsmapper/ClaimsFromTolkevaravApiTokenEnricher.java

```diff
@@ -0,0 +1,189 @@
+package ee.eki.tolkevarav.sso.keycloakserviceprovider.tokenclaimsmapper;
+
+import org.apache.http.client.utils.URIBuilder;
+import org.jboss.logging.Logger;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.models.ClientSessionContext;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ProtocolMapperModel;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.protocol.oidc.TokenManager;
+import org.keycloak.representations.IDToken;
+import org.keycloak.utils.StringUtil;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.time.Duration;
+
+import static java.net.http.HttpResponse.BodyHandlers.ofString;
+
+class ClaimsFromTolkevaravApiTokenEnricher {
+
+    private static final String SELECTED_INSTITUTION_ID_HEADER_KEY = "X-Selected-Institution-ID";
+    private static final String TARA_SUBJECT_CLAIM_ATTRIBUTE_KEY = "tara_subject_claim";
+    private final ConfigurationParameters configuration;
+    private final KeycloakSession keycloakSession;
+    private final UserSessionModel userSession;
+    private final ClientSessionContext clientSessionContext;
+
+    private static final Logger logger = Logger.getLogger(ClaimsFromTolkevaravApiTokenEnricher.class);
+
+    ClaimsFromTolkevaravApiTokenEnricher(ProtocolMapperModel mapperModel,
+                                                KeycloakSession keycloakSession,
+                                                UserSessionModel userSession,
+                                                ClientSessionContext clientSessionContext) {
+        this.configuration = ConfigurationParameters.fromModel(mapperModel);
+        this.keycloakSession = keycloakSession;
+        this.userSession = userSession;
+        this.clientSessionContext = clientSessionContext;
+    }
+
+    void enrichToken(IDToken token) {
+        try {
+            if (configuration.isInvalid()) {
+                throw new AcceptableMapperException("Mapper configuration is invalid: " + configuration);
+            }
+
+            var claims = queryClaimsFromApi(
+                retrievePersonalIdentificationCode(),
+                retrieveInstitutionId(),
+                generateAccessToken()
+            );
+
+            logger.infof("Claims received from Tõlkevärav API: %s. Adding it to claims in token.", claims);
+            token.getOtherClaims().put("tolkevarav", claims);
+        } catch (AcceptableMapperException exception) {
+            logger.info("Cancelled enriching of token because of an acceptable scenario: %s.", exception);
```

Comment:
That is a very good point! The front-end relies on the `tolkevarav` claims being populated and if we "silently" leave it empty here, then it might lead to some hard-to-track-down issues later. I will add a configuration parameter for whether `UnacceptableMapperException`s are swallowed or cause failure.

Reply:
Fixed with revision [ec3efd0](https://github.com/keeleinstituut/tv-sso/pull/5/commits/ec3efd0a125e6a62e31e004251d885b30e8feae3)